pkg/driver/controller.go

```diff
@@ -25,28 +26,298 @@ import (
+				return nil, status.Error(codes.InvalidArgument, "Error retrieving snapshot from the volumeContentSource")
+			}
+
+			id := sourceSnapshot.GetSnapshotId()
+			snapshot, err := d.cloud.DescribeSnapshot(ctx, id)
+			if err != nil {
+				return nil, err
+			}
+			snapshotArn = &snapshot.ResourceARN
+		}
+
+		volOptions := cloud.VolumeOptions{
+			Name:                          &volName,
+			StorageCapacityQuotaGiB:       storageCapacity,
+			StorageCapacityReservationGiB: storageCapacity,
+			SnapshotARN:                   snapshotArn,
+		}
+
+		for key, value := range volumeParams {
+			switch key {
+			case volumeParamsCopyTagsToSnapshots:
+				boolVal, err := util.StringToBoolPointer(value)
+				if err != nil {
+					return nil, status.Errorf(codes.InvalidArgument, "invalid parameter of %s: %s", volumeParamsCopyTagsToSnapshots, err)
+				}
+				volOptions.CopyTagsToSnapshots = boolVal
+			case volumeParamsDataCompressionType:
+				volOptions.DataCompressionType = util.StringToStringPointer(value)
+			case volumeParamsNfsExports:
+				volOptions.NfsExports = util.StringToStringPointer(value)
+			case volumeParamsParentVolumeId:
+				volOptions.ParentVolumeId = util.StringToStringPointer(value)
+			case volumeParamsReadOnly:
+				boolVal, err := util.StringToBoolPointer(value)
+				if err != nil {
+					return nil, status.Errorf(codes.InvalidArgument, "invalid parameter of %s: %s", volumeParamsReadOnly, err)
+				}
+				volOptions.ReadOnly = boolVal
+			case volumeParamsRecordSizeKiB:
+				i, err := util.StringToIntPointer(value)
+				if err != nil {
+					return nil, status.Errorf(codes.InvalidArgument, "invalid parameter of %s: %s", volumeParamsRecordSizeKiB, err)
+				}
+				volOptions.RecordSizeKiB = i
+			case volumeParamsUserAndGroupQuotas:
+				volOptions.UserAndGroupQuotas = util.StringToStringPointer(value)
+			case volumeParamsTags:
+				volOptions.Tags = util.StringToStringPointer(value)
+			}
+		}
+
+		v, err := d.cloud.CreateVolume(ctx, volName, volOptions)
+		if err != nil {
+			klog.V(4).Infof("CreateVolume error: ", err.Error())
+			switch {
+			case errors.Is(err, cloud.ErrInvalidParameter):
+				return nil, status.Error(codes.InvalidArgument, err.Error())
+			case errors.Is(err, cloud.ErrAlreadyExists):
+				return nil, status.Error(codes.AlreadyExists, err.Error())
+			default:
+				return nil, status.Errorf(codes.Internal, "Could not create volume %q: %v", volName, err)
+			}
+		}
+
+		err = d.cloud.WaitForVolumeAvailable(ctx, v.VolumeId)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Volume is not ready: %v", err)
+		}
+
+		fileSystem, err := d.cloud.DescribeFileSystem(ctx, v.FileSystemId)
+		if err != nil {
+			return nil, err
+		}
+
+		return &csi.CreateVolumeResponse{
+			Volume: &csi.Volume{
+				CapacityBytes: util.GiBToBytes(v.StorageCapacityReservationGiB),
+				VolumeId:      v.VolumeId,
+				VolumeContext: map[string]string{
+					volumeContextDnsName:    fileSystem.DnsName,
```

Comment:
We'd discussed potentially storing skipFinalBackup in the volume context, just in case we are able to retrieve in the DeleteVolume RPC in the future. Is there a reason why you didn't end up implementing it like that?

Reply:
This line points to the VolumeContext when creating a FSx Volume. On line 194 I add the `skipFinalBackup` context when creating a FSx Filesystem. skipFinalBackup only applies at a filesystem level which is why I left it out here.